packages/wdio-browserstack-service/src/instrumentation/performance/performance-tester.ts

```diff
@@ -0,0 +1,329 @@
+    private static _events: PerformanceEntry[] = []
+    private static _measuredEvents: PerformanceEntry[] = []
+    static started = false
+    static details: {[key: string]: PerformanceDetails} = {}
+    static eventsMap: {[key: string]: number} = {}
+    static browser?: WebdriverIO.Browser
+    static scenarioThatRan: string[]
+    static jsonReportDirName = 'performance-report'
+    static jsonReportDirPath = path.join(process.cwd(), 'logs', this.jsonReportDirName)
+    static jsonReportFileName = `${this.jsonReportDirPath}/performance-report-${PerformanceTester.getProcessId()}.json`
+
+    static startMonitoring(csvName: string = 'performance-report.csv') {
+        // Create performance-report dir if not exists already
+        if (!fs.existsSync(this.jsonReportDirPath)) {
+            fs.mkdirSync(this.jsonReportDirPath, { recursive: true })
+        }
+        this._observer = new PerformanceObserver(list => {
+            list.getEntries().forEach(entry => {
+                let finalEntry = entry
+                if (entry.entryType === 'measure') {
+                    finalEntry = entry.toJSON()
+                    if (this.details[entry.name]) {
+                        finalEntry = Object.assign(finalEntry, this.details[entry.name])
+                    }
+
+                    delete this.details[entry.name]
+                    this._measuredEvents.push(finalEntry)
+                } else if (process.env[PERF_MEASUREMENT_ENV]) {
+                    this._events.push(finalEntry)
+                }
+            })
+        })
+        const entryTypes: EntryType[] = ['measure']
+        if (process.env[PERF_MEASUREMENT_ENV]) {
+            entryTypes.push('function')
+        }
+        this._observer.observe({ buffered: true, entryTypes: entryTypes })
+        this.started = true
+        if (process.env[PERF_MEASUREMENT_ENV]) {
+            this._csvWriter = createObjectCsvWriter({
+                path: csvName,
+                header: [
+                    { id: 'name', title: 'Function Name' },
+                    { id: 'time', title: 'Execution Time (ms)' }
+                ]
+            })
+        }
+    }
+
+    static getPerformance() {
+        return performance
+    }
+
+    static calculateTimes(methods: string[]) : number {
+        const times: { [key: string]: number } = {}
+        this._events.map((entry) => {
+            if (!times[entry.name]) {
+                times[entry.name] = 0
+            }
+            times[entry.name] += entry.duration
+        })
+        const timeTaken = methods.reduce((a, c) => {
+            return times[c] + (a || 0)
+        }, 0)
+        BStackLogger.debug(`Time for ${methods} is ${timeTaken}`)
+        return timeTaken
+    }
+
+    static async stopAndGenerate(filename: string = 'performance-own.html') {
+        if (!this.started) {return}
+        await PerformanceTester.sleep(2000) // Wait to 2s just to finish any running callbacks for timerify
```

Comment:
What is timerify?

Reply:
This is used [here](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-browserstack-service/src/util.ts#L168). Not required if PERF_MEASUREMENT_ENV is false. Moved down